### DISCUSSION TITLE
How to change type ?

### DISCUSSION THREAD

**Discussion by michaelnabil230** [rebing/graphql-laravel#854] [Q&A]
Question:
Is there any way to switch to paginating products field or not?
the switch boolean flag will be for example called paginate which will be a query parameter

Versions:
graphql-laravel Version: 7.2
Laravel Version: 8.61
PHP Version: 8

```php
class ProductsQuery extends Query
{
    protected $attributes = [
        'name' => 'products',
    ];

    public function type(): Type
    {
        return Type::nonNull(new UnionType([
            'name' => 'ProductMe',
            'types' => [
                GraphQL::paginate('Product'),
                GraphQL::type('Product'),
            ],
            'resolveType' => function ($value) {
                // I Want kenwo if the resolve function return data with paginate or array of products for set my type
                // or anther function for chuck for the result 
                if ($value == '') {
                    return Type::listOf(GraphQL::type('Product'));
                }
                return GraphQL::paginate('Product');
            }
        ]));

        // return Type::nonNull(GraphQL::paginate('Product'));
    }

    public function args(): array
    {
        return [
            // start paginate
            'paginate' => [
                'name' => 'paginate',
                'type' => Type::boolean(),
                'rules' => ['boolean'],
                'defaultValue' => true,
            ],
            'limit' => [
                'name' => 'limit',
                'type' => Type::int(),
                'rules' => ['required_if:paginate,true'],
                'defaultValue' => 15,
            ],
            'page' => [
                'name' => 'page',
                'type' => Type::int(),
                'rules' => ['required_if:paginate,true'],
            ],
            // end paginate
        ];
    }

    public function resolve($root, array $args, $context, ResolveInfo $info, SelectFields $fields)
    {
        $select = $fields->getSelect();
        $with = $fields->getRelations();
        
        return Product::query()
            ->select($select)
            ->with($with)
            ->when($args['paginate'], function ($q) use ($args) {
                return $q->paginate($args['limit'], ['*'], 'page', $args['page']);
            }, function ($q) {
                return $q->get();
            });
    }
}
?>

**Comment by mfn** [comment]
Regarding the union type: I have not had this particular use-case, but on paper, yes, something like this should work.

- `types()` announces the possible unions
- `resovleType()` receives the `$value` and, based on its content, you determine/return the _actual_ type
  I've had cases where it was problematic just by inspection `$value` to know what type I would need, but that does not seems so in your case

Now whether this all can work that way in combination with the query builder, I can't say.

Is there a specific issue you have with your approach? Anything in particular not working?

**Reply by michaelnabil230** [reply]
This is the problem now and the code 

![image](https://user-images.githubusercontent.com/46572405/141464743-84fe011e-9cf9-4314-850a-17dadb5cb006.png)
![image](https://user-images.githubusercontent.com/46572405/141464994-21f9ab17-1f26-42ed-aa88-a9a44e41305b.png)

**Reply by michaelnabil230** [reply]
Please help and please explain to me because I am weak in English

**Reply by michaelnabil230** [reply]
And the `function resolveType` don't work It does not print anything in the log and also if you delete it, it does not miss anything

**Reply by mfn** [reply]
I've never used the `UnionType` raw inline like you did with this library; how did you came about that this is supported (note: maybe it is, but I never tried).

My point is: the readme does have (sparse, but it's there) example: https://github.com/rebing/graphql-laravel#unions

I suggest to follow this first before getting creative.

HTH

**Reply by michaelnabil230** [reply]
Its work but if changed `GraphQL::type('Product')` to this `Type::listOf(GraphQL::type('Product'))`  not work

```php
<?php

namespace App\GraphQL\Unions;

use GraphQL\Type\Definition\Type;
use Rebing\GraphQL\Support\UnionType;
use App\GraphQL\Types\Product\ProductType;
use Rebing\GraphQL\Support\Facades\GraphQL;

class ProductPaginateOrList extends UnionType
{
    protected $attributes = [
        'name' => 'ProductPaginateOrList',
        'description' => 'An example union',
    ];

    public function types(): array
    {
        return [
            GraphQL::paginate('Product'),
            GraphQL::type('Product')
        ];
    }

    public function resolveType($root)
    {
        if ($root instanceof ProductType) {
            return GraphQL::type('Product');
        }
        return GraphQL::paginate('Product');
    }
}
```

**Reply by michaelnabil230** [reply]
this the bug 
![Capture](https://user-images.githubusercontent.com/46572405/141534399-ed0958aa-e71c-42ae-9f89-0636aaf164fd.PNG)

**Reply by mfn** [reply]
I think what you're experiencing is similar to this https://github.com/graphql-python/graphene/issues/835#issuecomment-422321346

A union must be discrete types, you can't return `[Product] | Product` it seems. You would need an additional abstraction here, like `ProductResult | Product` and make `[Product]` part of `ProductResult`.

Or you rethink your strategy and just use two different queries; it's not uncommon to use a `productQuery` returning `product` and a `productsQuery` returning `[product]`

**Reply by michaelnabil230** [reply]
I did that but it didn't work well in query @mfn

**Reply by michaelnabil230** [reply]
I don't want to do more than query until the frontend programmer does a good job

**Reply by mfn** [reply]
> it didn't work

Sorry, but with that information alone I really can't help.

**Reply by michaelnabil230** [reply]
I did the first option (the `productResult` one). But the query that is been done on the database. But it showed a problem of type n+1

And unfortunately can't do the second option because it's going to be a ton of work for the frontend developer

**Reply by michaelnabil230** [reply]
https://github.com/rebing/graphql-laravel/blob/5817033018161eeab1092d2ae7293fded4bdebbd/src/Support/SelectFields.php#L256-L260

My problem here is that it executes the qualifier `['*']` and this disables the with or select functions and makes the n+1 problem that the query decides if I have a relationship with another table.
It also disables the GraphQL properties that reduce query

**Reply by michaelnabil230** [reply]
![image](https://user-images.githubusercontent.com/46572405/141678171-eb7ebb7d-2c0e-4497-8c50-dbc4c269468b.png)
![image](https://user-images.githubusercontent.com/46572405/141678297-5d9457c8-ba96-458b-b5f6-6377d173c892.png)

As shown in the picture here, the problem is in the query


```php
    public function resolve($root, array $args, $context, ResolveInfo $info, SelectFields $fields)
    {
        $select = $fields->getSelect();
        $with = $fields->getRelations();
        
        return Product::query()
            ->select($select)
            ->with($with)
            ->when($args['paginate'], function ($q) use ($args) {
                return $q->paginate($args['limit'], ['*'], 'page', $args['page']);
            }, function ($q) {
                return $q->get();
            });
    }

**Reply by michaelnabil230** [reply]
Hi friend are you here @mfn

**Reply by mfn** [reply]
Sorry, I don't have further suggestion currently. I'm not a user of `SelectFields`. I exclusively use solutions like https://www.google.com/search?hl=en&q=graphql%20dataloader%20php .

**Reply by michaelnabil230** [reply]
But i'm used `with` in `Eloquent` model in laravel maybe don't need another package, right ?

**Reply by michaelnabil230** [reply]
And maybe the my problem here 
https://github.com/rebing/graphql-laravel/blob/5817033018161eeab1092d2ae7293fded4bdebbd/src/Support/SelectFields.php#L256-L260

**Comment by mehrabx** [comment]
I have the same problem. When we use ‍``UnionType``, there is a potential problem of n+1.

https://github.com/rebing/graphql-laravel/blob/5817033018161eeab1092d2ae7293fded4bdebbd/src/Support/SelectFields.php#L256-L260

Unlike line 258 of code, we know which fields are required according to the types defined in the union class.
